src/services/TestResultsService.ts

```diff
@@ -579,121 +586,56 @@ export class TestResultsService {
    *
    * @param payload
    */
-  public generateExpiryDate(payload: ITestResultPayload) {
+  public async generateExpiryDate(payload: ITestResultPayload): Promise<ITestResultPayload> {
     moment.tz.setDefault("UTC");
-    if (payload.testStatus !== TEST_STATUS.SUBMITTED) {
+    try {
+      if (payload.testStatus !== TEST_STATUS.SUBMITTED ||
+          this.isNotAllowedVehicleTypeForExpiry(payload.vehicleType)) {
+        return Promise.resolve(payload);
+      }
+      const expiryTestTypes = payload.testTypes.filter((testType) => this.isAllowedTestTypeForExpiry(testType));
+
+      const recentExpiry =  await this.vehicleTestController.dataProvider.getMostRecentExpiryDate(payload.systemNumber);
+
+      expiryTestTypes.forEach((testType: any, index: number) => {
```

Comment:
up to you but we could be chaining the forEach() after the filter.
I will have a question for this function, so I would love to have a chat about that whenever you are available please.